Fixes for Windows

djdv opened the pull request:
Continuation on: https://github.com/toppair/peek.nvim/issues/14 based on @yehuohan's finding.
yehuohan, if you don't mind, can you see if this general path resolution works for you without needing to hard code the vim path?

Notes:
- I have not worked with Typescript before. This works for me but I do not know if it will break on other platforms.
  - Please test/advise. :^) I wonder if the calls can be simplified, or if this is an okay conversion.
- I needed to remove the signals as suggested too https://github.com/toppair/peek.nvim/issues/14#issuecomment-1465417029
  - This part is not included in the PR because I don't know how to branch for OS specific code in Deno, but it should be as easy as using different sets for different operating systems. If you let me know I can push it, or someone else can.
- I am using `lazy.nvim` and using the property `build = 'deno task --quiet build:fast'`, but it does not run automatically. However, I think this is a bug in `lazy.nvim` or my configuration. Running it manually builds okay before and after patch.

Here it is running on my machine before and after the patch is applied: https://www.youtube.com/watch?v=ao0d3cWNn-A

jakehl commented:
I don't think we should need system native path names here. We're in a web context and `file://...` is perfectly cross platform in the web. Which part of #14 is this aiming to fix?

As for the signals, you can use `Deno.build.os` to get the operating system. We just need to supply a different list of signals if it's windows.

jakehl commented on app/src/main.ts:
```diff
@@ -173,15 +173,21 @@ async function init(socket: WebSocket) {
+if (Deno.build.os === "windows") {
+	signals = ['SIGINT', 'SIGBREAK'];
+} else {
+	signals = [
```
nitpick: I think it might be more readable if you keep signals as a const and assign it with a ternary operator. 

Something like:
```ts
const win_signals = ['SIGINT', 'SIGBREAK'];
const unix_signals = ['SIGINT', 'SIGUSR2', '...etc'];
const signals = Deno.build.os === 'windows' ? win_signals : unix_signals;
```

This way it's immediately clear that there's a condition that dictates the value of signals when it's declared rather than it being assigned after the fact.

djdv replied:
Done. 🙏
I had to change the `for` loop declarations though.
Trying to assert as a `const` was throwing an assertion error, and trying to `as` with `Deno.Signal` was giving another error.
Assuming that's expected / required here since `signals` itself is variable instead of a constant literal now.

jakehl replied:
Yeah, using let there should be fine. We've added `as const` to the individual arrays so we should still get type safety inside the loop as both `win_signals` and `unix_signals` satisfy the `Deno.Signal` type constraint. I'd expect to see an error in `Deno.addSignalListener` if it wasn't valid. Great work!

djdv commented:
Thanks Jake. I looked it over and changed some things.

>We're in a web context and file://... is perfectly cross platform in the web. Which part of https://github.com/toppair/peek.nvim/issues/14 is this aiming to fix?

Without the change to `Deno.run`'s `cwd` parameter, I encounter the same issue as in #14.
```
Peek error: error: Uncaught (in promise) Error: The directory name is invalid. (os error 267)
        const webview = Deno.run({
```
`Deno.run` seems to be related to OS subprocess management.
My host (Windows) refuses to invoke `run`'s `cmd` when the path contained in `cwd` is not a valid path.
@yehuohan hardcoded a Windows path to work around this, and I'm trying to generalize that with this change: https://github.com/toppair/peek.nvim/pull/27/commits/811724b35311ec4965119efde16f5b6c7c20db89

`Deno.open` seems to accept either a path or a URL, so a URL.
This seems to work without a conversion (at least on my system which accepts both `/` and `\` path separators), so I've remove my changes around that.
Originally, I missed the part where it was being converted back to a URL anyway, so I don't think the patch even affected the call to `open`. Or if it did, was converting from URL -> path -> URL (as a mistake).

---

I added a commit to handle separate signals: https://github.com/toppair/peek.nvim/pull/27/commits/da11d92ee997749aafe4ec13eac51b2c3469eff1
But I don't typically write JavaScript or Typescript so I'm not sure what the best form to express this is.
Does this seem right?

jakehl commented:
No worries! Ah I get you, I missed the `cwd` part. Good catch! Yeah it'll definitely need to be more along the lines of your fix then. I'll check out the changes in that area.

I Added a comment on the signals part, but it's honestly a really minor (and opinionated) nitpick, which is ultimately up to @toppair to decide on 😆 so don't worry too much about that. Your approach is perfectly fine in javascript/typescript land!

jakehl commented on app/src/main.ts:
```diff
@@ -75,7 +75,7 @@ async function init(socket: WebSocket) {
   if (app === 'webview') {
     const webview = Deno.run({
-      cwd: dirname(new URL(Deno.mainModule).pathname),
+      cwd: dirname(fromFileUrl(new URL(Deno.mainModule))),
```
I think you should be able to do just `dirname(fromFileUrl(Deno.mainModule))` here, no need to do the `new URL()`. I can't test this on windows, though.

Otherwise looks great!

djdv replied:
Thanks again.
Yeah this seems to build and run as expected on my Windows machine. I force pushed over it as well.

yehuohan commented:
@djdv Great work. I can preview markdown in on windows with your PR without any issue.

toppair commented:
@djdv @jakehl @yehuohan Thank you :pray:
